Document Kubernetes Kit backend session expiration policy (#4144)
vaadin/docs: articles/tools/kubernetes/configuration.adoc

    host: redis-service
----
--

== Backend Session Expiration Policy

> Backend session expiration policy allows to define an expiration timeout for the backend stored session. By default no expiration is set, but can be expiration can be activated setting the property or by providing a custom SessionExpirationPolicy bean. If no unit is provided in the property, seconds are assumed.

Reviewer: ```suggestion
The backend session expiration policy allows you to define an expiration timeout for the backend stored session. By default no expiration is set, but expiration can be activated by setting the property or by providing a custom `SessionExpirationPolicy` bean. If no unit is provided in the property, seconds are assumed.
```

> Backend session expiration policy allows to define an expiration timeout for the backend stored session. By default no expiration is set, but can be expiration can be activated setting the property or by providing a custom SessionExpirationPolicy bean. If no unit is provided in the property, seconds are assumed.

Reviewer 2: By implementing the `SessionExpirationPolicy` interface, the developer can define the expiration timeout.
However, the property does not define the timeout value, but the amount of time to be added to the HTTP session timeout to determine the expiration of the backend session. If time unit is not given, milliseconds is assumed.

Reviewer 2: It would be nice to add also a `SessionExpirationPolicy` implementation example

Author: Thanks for clarification

Author: Should this suffice for code example?

        @Bean
        SessionExpirationPolicy sessionExpirationPolicy() {
            return Duration.ofMinutes(60);
        }

Reviewer 2: The `SessionExpirationPolicy.apply()` method takes current HTTP session timeout as argument
```
@Bean
SessionExpirationPolicy sessionExpirationPolicy() {
    return sessionTimeout -> Duration.ofMinutes(60);
}
```

Reviewer: > If time unit is not given, milliseconds is assumed.

This may be the default, but it's not particularly useful. If seconds are more appropriate, perhaps the property should be annotated with `@DurationUnit(ChronoUnit.SECONDS)` and the original phrasing kept?

Reviewer 2: Sound a good idea

    host: redis-service
----
--

== Backend Session Expiration Policy

> The backend session expiration policy allows you to define an expiration timeout for the backend stored session. By default no expiration is set, but expiration can be activated by setting the property or by providing a custom `SessionExpirationPolicy` bean. The property defines the amount of time to be added to the HTTP session timeout to determine the expiration of the backend session. If time unit is not given, milliseconds is assumed.

Reviewer 2: ```suggestion
The backend session expiration policy allows you to define an expiration timeout for the backend stored session. By default no expiration is set, but expiration can be activated by setting the `vaadin.kubernetes.backend-session-expiration-tolerance` property or by providing a custom `SessionExpirationPolicy` bean. The property defines the amount of time to be added to the HTTP session timeout to determine the expiration of the backend session. If time unit is not given, milliseconds is assumed.
```

[.example]
--
.application.properties
[source,properties]
----

> vaadin.kubernetes.backend-session-expiration-tolerance: duration

Reviewer 2: I would add a proper value in the example

```suggestion
# Given an HTTP session timeout of 30 minuts, the session stored in the serialization backed will expires 35 minutes of inactivity
vaadin.kubernetes.backend-session-expiration-tolerance: 5m
```

    host: redis-service
----
--

== Backend Session Expiration Policy

> The backend session expiration policy allows you to define an expiration timeout for the backend stored session. By default, no expiration is set. However, expiration can be activated by setting the `vaadin.kubernetes.backend-session-expiration-tolerance` property, or by providing a custom `SessionExpirationPolicy` bean. The property defines the amount of time to be added to the HTTP session timeout to determine the expiration of the backend session. If a time unit is not given, milliseconds is assumed.

Reviewer 3: Am i correct in understanding from this that the property and SessionExpirationPolicy bean work differently? The property configures a duration that is added to the http session timeout to determine expiration, whereas the bean returns a duration which is taken as-is i.e. not added to the http session timeout?

Reviewer 2: Yes, exactly. The default policy is to not have backend expiration, but if the property is set, the Kit will activate a strategy based on the HTTP session timeout, adjusted by the given amount of time.
However, if you provide a custom implementation, you can apply whatever algorithm your application may need,

Reviewer 2: Keep in mind that the `SessionExpirationPolicy` implementation gets the HTTP session timeout as input parameter, so they work differently in the sense that implementing the bean you are free to decide how long the backed session would be retained, but you can still take the decision based on the application HTTP session timeout.

Reviewer 3: ok understood, thanks for the clarification !

[.example]
--
.application.properties
[source,properties]
----

> # Given an HTTP session timeout of 30 minutes, the session stored in the backed will expire after 35 minutes of inactivity

Reviewer 3: ```suggestion
# Given an HTTP session timeout of 30 minutes, the session stored in the backend will expire after 35 minutes of inactivity
```

vaadin.kubernetes.backend-session-expiration-tolerance: 5m
----

.application.yaml
[source,yaml]
----
vaadin:
  kubernetes:

>     backend-session-expiration-tolerance: duration

Reviewer 2: ```suggestion
    backend-session-expiration-tolerance: 5m
```

>     backend-session-expiration-tolerance: duration

Reviewer 2: Maybe better use the same real value also in this example.
```suggestion
    backend-session-expiration-tolerance: 5m
```

> [.example]
> --
> .application.properties
> [source,properties]
> ----
> # Given an HTTP session timeout of 30 minuts, the session stored in the backed will expire 35 minutes of inactivity
> vaadin.kubernetes.backend-session-expiration-tolerance: 5m
> ----
> 
> .application.yaml
> [source,yaml]
> ----
> vaadin:
>   kubernetes:
>     backend-session-expiration-tolerance: duration
> ----
> 
> .Java
> [source,java]
> ----
> @Bean
> SessionExpirationPolicy sessionExpirationPolicy() {
>     return sessionTimeout -> Duration.ofMinutes(60);
> }
> ----
> --

Reviewer: Improved layout for example and fixed typos:
```suggestion
[.example]
--
Given an HTTP session timeout of 30 minutes, the session stored in the backend will expire 35 minutes of inactivity.
[source,properties]
----
<source-info group="application.properties"></source-info>
vaadin.kubernetes.backend-session-expiration-tolerance: 5m
----

Given an HTTP session timeout of 30 minutes, the session stored in the backend will expire 35 minutes of inactivity.
[source,yaml]
----
<source-info group="application.yaml"></source-info>
vaadin:
  kubernetes:
    backend-session-expiration-tolerance: 5m
----

The session stored in the backend will expire 35 minutes of inactivity, regardless of the HTTP session timeout.
[source,java]
----
<source-info group="Java bean"></source-info>
@Bean
SessionExpirationPolicy sessionExpirationPolicy() {
    return sessionTimeout -> Duration.ofMinutes(35);
}
----
--
```
